### ISSUE TITLE
create-dirs-from-rpmdb.service enabled and always failed

### ISSUE THREAD

**Issue by Ciiss0u** [openSUSE/transactional-update#97]
Hi,

Not sure I'm at the right place for opening this issue;
I'm new to openSUSE, and running Tumbleweed in transactional-server mode.

I see that the create-dirs-from-rpmdb.service is always failed, because of the readonly fs.

I ran it in a transactional-update shell and it worked:
![create RPM directoris in snapshot](https://i.imgur.com/uK2mpPF.png)

I don't know if this service should be enabled or not on a transactional server, but in fact it will still not work as-is, outside of the transactional-update process..
![failed service create dirs from RPMDB](https://i.imgur.com/SWysYvy.png)

It seems implemented in the transactional-update process already ? (https://github.com/openSUSE/transactional-update/tree/master/sbin), but I am not sure why the service is enabled and why it has created missing folders when ran in a snapshot... Like if it was never called during transactional-update ?

**Comment by laenion** [comment]
### Some Background

`create_dirs_from_rpmdb` is not supposed to be run from inside the transactional-update environment - let's have a look at what it is actually doing instead before looking at the problem:
* On a transactional system only the root file system is read-only (so basically mostly _/usr_ and _/boot_), but several other locations have to be writable. Among these are _/etc_ and - the relevant part for `create_dirs_from_rpmdb` - _/var_. You can check all the writable subvolumes with `findmnt`.
* One important part of the transactional update is that the update is performed in a separate snapshot in the background - so if the update fails the currently running system hasn't been influenced.
* This however means that _/var_ can neither be snapshotted nor mounted into the transactional-update environment - the directory contains production data such as databases or containers. Snapshots would discard new data, and mounting the directory into the update environment would mean that scripts may alter the files in there. This would be fatal if an update script would perform a database migration or similar interruptive things, as the currently running system probably doesn't know how to handle these things.
* Due to that */var* is **not** mounted into the update environment.

However some applications may still want to create directories in there. This is where the `create_dirs_from_rpmdb` service comes into play: It will check whether the RPM database has been changed since the last boot, and if so it will create all the missing directories early during boot, before any of the packages requiring these directories will start. This is the whole purpose of the service.

### The Problem

So now let's get to the actual bug report: Why does the service fail?

This seems to be caused by the package `pcp-conf`: It is owning/creating the directory _/var/lib/pcp/config/derived_, but doesn't own the directory _/var/lib/pcp/config_. I think this is a packaging error in `pcp-conf`, but I'll verify this tomorrow.

**Comment by Vogtinator** [comment]
It's not just `pcp-conf`, there's also the independent `Failed to resolve uuidd/uuidd`, which is [boo#1206690](https://bugzilla.opensuse.org/show_bug.cgi?id=1206690).

**Comment by Ciiss0u** [comment]
Hi,

Thanks for the explanations, I now well understand the purpose of this service. I have reverted the changes I did on my system when ran the service in a transactional-update snapshot.

For the issue with pcp-conf, I do not have any /var/lib/pcp folder in my running system (in whole /var mounted from the brtfs subvolume). 

Not sure it can help for debug but I figured that /var/lib/pcp/config/derived is present in the transactional-updates snapshots (so when var is not mounted from the btrfs subvolume).
Here's a screenshot from one of my oldest snapshot (so before I ran the create dirs command in transactional-update, so these folders were not created by this create_dirs... command but probably when the package was installed/updated. And successfully created.)
![pcp folders present](https://i.imgur.com/G2MPcnM.png)

After I ran the create_dirs_from_rpmdb -v  command in transactional-update:
![after create_dirs](https://i.imgur.com/o4ofG5p.png)

On my current booted snapshot, after I did some cleanup:
![cleanup](https://i.imgur.com/FagaG25.png)

Also I'm not sure but maybe /var should be read-only in transactional update environment or removed/recreated for each snapshot before mounting the needed subvolume's folders in the transactional update environment ?  To not having it filling up with useless files when installing / updating packages...
